pkg/worker/verifier/transactional_producer_worker.go

```diff
@@ -0,0 +1,345 @@
+		}
+
+		if n <= 0 {
+			return nil
+		} else {
+			// Record that we took another run at produceInner
+			pw.Status.Restarts += 1
+		}
+	}
+}
+
+func (pw *TransactionalProducerWorker) produceInner(n int64) (int64, []BadOffset, error) {
+	opts := pw.config.workerCfg.MakeKgoOpts()
+	randId := uuid.New()
+
+	opts = append(opts, []kgo.Opt{
+		kgo.ProducerBatchCompression(kgo.NoCompression()),
+		kgo.RequiredAcks(kgo.AllISRAcks()),
+		kgo.RecordPartitioner(kgo.ManualPartitioner()),
+		kgo.TransactionalID("p" + randId.String()),
+		kgo.TransactionTimeout(2 * time.Minute),
+	}...)
+	client, err := kgo.NewClient(opts...)
+	if err != nil {
+		log.Errorf("Error creating Kafka client: %v", err)
+		return 0, nil, err
+	}
+
+	currentOffsets := GetOffsets(client, pw.config.workerCfg.Topic, pw.config.nPartitions, -1)
+
+	for i, o := range currentOffsets {
+		log.Infof("Produce start offset %s/%d %d...", pw.config.workerCfg.Topic, i, o)
+	}
+
+	var wg sync.WaitGroup
+
+	errored := false
+	produced := int64(0)
+
+	// Channel must be >= concurrency
+	bad_offsets := make(chan BadOffset, 16384)
+	concurrent := semaphore.NewWeighted(4096)
+
+	log.Infof("Producing %d messages (%d bytes)", n, pw.config.messageSize)
+
+	if err := client.BeginTransaction(); err != nil {
+		log.Errorf("Couldn't start a transaction: %v", err)
+		pw.Status.FailedTransactions += 1
+		return 0, nil, err
+	}
+
+	// BeginTransaction will leave one control record in
+	// every partition's log.
+	for i, _ := range currentOffsets {
+		currentOffsets[i] += 1
+	}
+
+	willAbort := pw.config.abortRate >= rand.Float64()
+
+	for i := int64(0); i < n && len(bad_offsets) == 0; i = i + 1 {
+		concurrent.Acquire(context.Background(), 1)
+		produced += 1
+		pw.Status.Sent += 1
+		var p = rand.Int31n(pw.config.nPartitions)
+
+		if i > 0 && i%int64(pw.config.msgsPerTransaction) == 0 {
+			if err := client.Flush(context.Background()); err != nil {
+				log.Errorf("Unable to flush: %v", err)
+				errored = true
+				pw.Status.FailedTransactions += 1
+				break
+			}
+			if err := client.EndTransaction(context.Background(), kgo.TransactionEndTry(!willAbort)); err != nil {
+				log.Errorf("unable to end transaction: %v", err)
+				errored = true
+				pw.Status.FailedTransactions += 1
+				break
+			}
+			if err := client.BeginTransaction(); err != nil {
+				log.Errorf("Couldn't start a transaction: %v", err)
+				errored = true
+				pw.Status.FailedTransactions += 1
+				break
+			}
+
+			for i, _ := range currentOffsets {
+				if !willAbort {
+					// EndTransaction and BeginTransaction will each leave
+					// one control record in each partition's log.
+					currentOffsets[i] += 2
+				} else {
+					// In this case a transaction that was just aborted.
+					// A abort doesn't leave a control record currently
+					// So just account for the control record left by BeginTransaction
+					// TODO: this behavior has changed on the tip of dev and aborts
+					// now leave a control record
```

Comment:
Aborts always had a control record in the log. "prepare" batch was omitted in the latest round of changes that went in the last couple of weeks.

Before:
```
begin/fence
data0
data1
...
prepare
commit/abort
```

now.

```
begin
data0
data1
...
commit/abort
```